File: neonlink_app/lib/data/models/telemetry_data.dart
import 'package:json_annotation/json_annotation.dart';

part 'telemetry_data.g.dart';

/// Главный класс телеметрии, получаемый от сервера
/// Версия API: 1.0.0
@JsonSerializable()
class TelemetryData {
  /// Версия API для совместимости клиента и сервера
  final String version;
  
  final int timestamp;
  final SystemInfo system;
  final GamingInfo? gaming;
  final String adminLevel;

  TelemetryData({
    this.version = '1.0.0',
    required this.timestamp,
    required this.system,
    this.gaming,
    this.adminLevel = 'Full',
  });

  factory TelemetryData.fromJson(Map<String, dynamic> json) =>
      _$TelemetryDataFromJson(json);

  Map<String, dynamic> toJson() => _$TelemetryDataToJson(this);
}

/// Информация о системе
@JsonSerializable()
class SystemInfo {
  final CpuInfo cpu;
  final GpuInfo gpu;
  final RamInfo ram;
  final List<StorageInfo> storage;
  final NetworkInfo? network;

  SystemInfo({
    required this.cpu,
    required this.gpu,
    required this.ram,
    required this.storage,
    this.network,
  });

  factory SystemInfo.fromJson(Map<String, dynamic> json) =>
      _$SystemInfoFromJson(json);

  Map<String, dynamic> toJson() => _$SystemInfoToJson(this);
}

/// Информация о CPU
@JsonSerializable()
class CpuInfo {
  final String name;
  final double usage;
  final double temp;
  final double clock;
  final double? power;
  final List<CpuCoreInfo> cores;

  CpuInfo({
    this.name = '',
    this.usage = 0,
    this.temp = 0,
    this.clock = 0,
    this.power,
    this.cores = const [],
  });

  factory CpuInfo.fromJson(Map<String, dynamic> json) =>
      _$CpuInfoFromJson(json);

  Map<String, dynamic> toJson() => _$CpuInfoToJson(this);
}

/// Информация о ядре CPU
@JsonSerializable()
class CpuCoreInfo {
  final int id;
  final double usage;
  final double? temp;
  final double? clock;

  CpuCoreInfo({
    required this.id,
    this.usage = 0,
    this.temp,
    this.clock,
  });

  factory CpuCoreInfo.fromJson(Map<String, dynamic> json) =>
      _$CpuCoreInfoFromJson(json);

  Map<String, dynamic> toJson() => _$CpuCoreInfoToJson(this);
}

/// Информация о GPU
@JsonSerializable()
class GpuInfo {
  final String name;
  final String type;
  final double usage;
  final double temp;
  final double vramUsed;
  final double vramTotal;
  final double clock;
  final double? memoryClock;
  final double? power;
  final int? fanSpeed;

  GpuInfo({
    this.name = '',
    this.type = '',
    this.usage = 0,
    this.temp = 0,
    this.vramUsed = 0,
    this.vramTotal = 0,
    this.clock = 0,
    this.memoryClock,
    this.power,
    this.fanSpeed,
  });

  factory GpuInfo.fromJson(Map<String, dynamic> json) =>
      _$GpuInfoFromJson(json);

  Map<String, dynamic> toJson() => _$GpuInfoToJson(this);

  /// Вычисляемое поле: процент использования VRAM
  double get vramUsagePercent => 
      vramTotal > 0 ? (vramUsed / vramTotal) * 100 : 0;
}

/// Информация о RAM
@JsonSerializable()
class RamInfo {
  final double used;
  final double total;
  final int? speed;

  RamInfo({
    this.used = 0,
    this.total = 0,
    this.speed,
  });

  factory RamInfo.fromJson(Map<String, dynamic> json) =>
      _$RamInfoFromJson(json);

  Map<String, dynamic> toJson() => _$RamInfoToJson(this);

  /// Вычисляемое поле: доступная память
  double get available => total - used;

  /// Вычисляемое поле: процент использования
  double get usedPercent => total > 0 ? (used / total) * 100 : 0;
}

/// Информация о накопителе
@JsonSerializable()
class StorageInfo {
  final String name;
  final double? temp;
  final int? health;
  final StorageSmartData? smart;

  StorageInfo({
    this.name = '',
    this.temp,
    this.health,
    this.smart,
  });

  factory StorageInfo.fromJson(Map<String, dynamic> json) =>
      _$StorageInfoFromJson(json);

  Map<String, dynamic> toJson() => _$StorageInfoToJson(this);
}

/// SMART данные накопителя
@JsonSerializable()
class StorageSmartData {
  final int? tbw;
  final int? powerOnHours;
  final int? reallocatedSectors;
  final int? temperature;

  StorageSmartData({
    this.tbw,
    this.powerOnHours,
    this.reallocatedSectors,
    this.temperature,
  });

  factory StorageSmartData.fromJson(Map<String, dynamic> json) =>
      _$StorageSmartDataFromJson(json);

  Map<String, dynamic> toJson() => _$StorageSmartDataToJson(this);
}

/// Информация о сети
@JsonSerializable()
class NetworkInfo {
  final double download;
  final double upload;
  final int ping;
  final String? ip;

  NetworkInfo({
    this.download = 0,
    this.upload = 0,
    this.ping = 0,
    this.ip,
  });

  factory NetworkInfo.fromJson(Map<String, dynamic> json) =>
      _$NetworkInfoFromJson(json);

  Map<String, dynamic> toJson() => _$NetworkInfoToJson(this);
}

/// Информация о игровом режиме
@JsonSerializable()
class GamingInfo {
  final bool active;
  final int? fps;
  final int? fps1Low;
  final double? frametime;
  final String? activeProcess;

  GamingInfo({
    this.active = false,
    this.fps,
    this.fps1Low,
    this.frametime,
    this.activeProcess,
  });

  factory GamingInfo.fromJson(Map<String, dynamic> json) =>
      _$GamingInfoFromJson(json);

  Map<String, dynamic> toJson() => _$GamingInfoToJson(this);
}

File: src/NeonLink.Server/Models/TelemetryData.cs
using System.Text.Json.Serialization;

namespace NeonLink.Server.Models;

/// <summary>
///     Главный класс телеметрии, отправляемый на Android клиент
///     Версия API: 1.0.0
/// </summary>
public class TelemetryData
{
    /// <summary>
    ///     Версия API для совместимости клиента и сервера
    /// </summary>
    [JsonPropertyName("version")]
    public string Version { get; set; } = "1.0.0";

    [JsonPropertyName("timestamp")]
    public long Timestamp { get; set; }

    [JsonPropertyName("system")]
    public SystemInfo System { get; set; } = new();

    [JsonPropertyName("gaming")]
    public GamingInfo? Gaming { get; set; }

    [JsonPropertyName("adminLevel")]
    public string AdminLevel { get; set; } = "Full";
}

/// <summary>
///     Информация о системе
/// </summary>
public class SystemInfo
{
    [JsonPropertyName("cpu")]
    public CpuInfo Cpu { get; set; } = new();

    [JsonPropertyName("gpu")]
    public GpuInfo Gpu { get; set; } = new();

    [JsonPropertyName("ram")]
    public RamInfo Ram { get; set; } = new();

    [JsonPropertyName("storage")]
    public List<StorageInfo> Storage { get; set; } = new();

    [JsonPropertyName("network")]
    public NetworkInfo? Network { get; set; }
}

/// <summary>
///     Информация о CPU
/// </summary>
public class CpuInfo
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("usage")]
    public double Usage { get; set; }

    [JsonPropertyName("temp")]
    public double Temp { get; set; }

    [JsonPropertyName("cores")]
    public List<CpuCoreInfo> Cores { get; set; } = new();

    [JsonPropertyName("clock")]
    public double Clock { get; set; }

    [JsonPropertyName("power")]
    public double? Power { get; set; }
}

/// <summary>
///     Информация о ядре CPU
/// </summary>
public class CpuCoreInfo
{
    [JsonPropertyName("id")]
    public int Id { get; set; }

    [JsonPropertyName("usage")]
    public double Usage { get; set; }

    [JsonPropertyName("temp")]
    public double? Temp { get; set; }

    [JsonPropertyName("clock")]
    public double? Clock { get; set; }
}

/// <summary>
///     Информация о GPU
/// </summary>
public class GpuInfo
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty; // NVIDIA, AMD, Intel

    [JsonPropertyName("usage")]
    public double Usage { get; set; }

    [JsonPropertyName("temp")]
    public double Temp { get; set; }

    [JsonPropertyName("vramUsed")]
    public double VramUsed { get; set; }

    [JsonPropertyName("vramTotal")]
    public double VramTotal { get; set; }

    [JsonPropertyName("clock")]
    public double Clock { get; set; }

    [JsonPropertyName("memoryClock")]
    public double? MemoryClock { get; set; }

    [JsonPropertyName("power")]
    public double? Power { get; set; }

    [JsonPropertyName("fanSpeed")]
    public int? FanSpeed { get; set; }
}

/// <summary>
///     Информация о RAM
/// </summary>
public class RamInfo
{
    [JsonPropertyName("used")]
    public double Used { get; set; }

    [JsonPropertyName("total")]
    public double Total { get; set; }

    [JsonPropertyName("speed")]
    public int? Speed { get; set; }

    [JsonPropertyName("available")]
    public double Available => Total - Used;

    [JsonPropertyName("usedPercent")]
    public double UsedPercent => Total > 0 ? (Used / Total) * 100 : 0;
}

/// <summary>
///     Информация о накопителе
/// </summary>
public class StorageInfo
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("temp")]
    public double? Temp { get; set; }

    [JsonPropertyName("health")]
    public int? Health { get; set; }

    [JsonPropertyName("smart")]
    public StorageSmartData? Smart { get; set; }
}

/// <summary>
///     SMART данные накопителя
/// </summary>
public class StorageSmartData
{
    [JsonPropertyName("tbw")]
    public int? Tbw { get; set; } // Total Bytes Written

    [JsonPropertyName("powerOnHours")]
    public int? PowerOnHours { get; set; }

    [JsonPropertyName("reallocatedSectors")]
    public int? ReallocatedSectors { get; set; }

    [JsonPropertyName("temperature")]
    public int? Temperature { get; set; }
}

/// <summary>
///     Информация о сети
/// </summary>
public class NetworkInfo
{
    [JsonPropertyName("download")]
    public double Download { get; set; }

    [JsonPropertyName("upload")]
    public double Upload { get; set; }

    [JsonPropertyName("ping")]
    public int Ping { get; set; }

    [JsonPropertyName("ip")]
    public string? Ip { get; set; }
}

/// <summary>
///     Информация о игровом режиме
/// </summary>
public class GamingInfo
{
    [JsonPropertyName("active")]
    public bool IsActive { get; set; }

    [JsonPropertyName("fps")]
    public int? Fps { get; set; }

    [JsonPropertyName("fps1Low")]
    public int? Fps1Low { get; set; }

    [JsonPropertyName("frametime")]
    public double? Frametime { get; set; }

    [JsonPropertyName("activeProcess")]
    public string? ActiveProcess { get; set; }
}
